### PR TITLE
Refactor tests in test/dist_autograd_test.py file

### PR DESCRIPTION
**Pull request by fkhan1337**
Summary:
Refactored the following fairly similar functions:
  1. `test_context_cleanup_tensor_with_grad`
  2. `test_context_cleanup_tensor_no_grad`
  3. `test_context_cleanup_no_tensors`
by creating a helper function `context_cleanup_test_helper` that can be invoked with the appropriate arguments.

Test Plan: Verified by running tests. 

Differential Revision: D19269246

### CODE REVIEW

**Comment by rohan-varma** [comment]
Relevant issue: https://github.com/pytorch/pytorch/issues/31072.

File: test/dist_autograd_test.py
Comment (rohan-varma) [inline test/dist_autograd_test.py]: nit: use `camel_case` for `rpcargs`.

File: test/dist_autograd_test.py
Comment (rohan-varma) [inline test/dist_autograd_test.py]: nit: `ret` is not needed since we don't consume the return value of this

File: test/dist_autograd_test.py
Comment (rohan-varma) [inline test/dist_autograd_test.py]: Could you change this to not have a `t2` at all and just use `rpc_args=(t1, t1)`?

**Review by rohan-varma — COMMENTED** [review]
Thanks for working on this! Looks good overall, a few minor comments inline. Also, can we refactor `test_context_cleanup_nested_rpc` to use this helper function? It might be a bit tricker since that does a nested rpc. One idea is that we can have a `nested=True` flag that controls this logic (which would be False for all tests other than the nested test).

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 34091d85:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=34091d85d2e6582fcac9c33d1c58fce860ce4f06).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4111300) pytorch_linux_xenial_cuda9_cudnn7_py3_NO_AVX_NO_AVX2_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=79602625) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=79602625)) :snowflake:
<details>
<summary>
<code>Jan 03 00:55:39 AssertionError: 1024 not less than or equal to 1e-05 : __main__.TestAutogradDeviceTypeCUDA.test_logdet_1x1_cuda leaked 1024 bytes CUDA memory on device 0</code>
</summary>

```
Jan 03 00:55:39 ====================================================================== 
Jan 03 00:55:39 FAIL [0.097s]: test_logdet_1x1_cuda (__main__.TestAutogradDeviceTypeCUDA) 
Jan 03 00:55:39 ---------------------------------------------------------------------- 
Jan 03 00:55:39 Traceback (most recent call last): 
Jan 03 00:55:39   File "/var/lib/jenkins/workspace/test/common_utils.py", line 665, in wrapper 
Jan 03 00:55:39     method(*args, **kwargs) 
Jan 03 00:55:39   File "/var/lib/jenkins/workspace/test/common_utils.py", line 521, in __exit__ 
Jan 03 00:55:39     self.name, after - before, i)) 
Jan 03 00:55:39   File "/var/lib/jenkins/workspace/test/common_utils.py", line 877, in assertEqual 
Jan 03 00:55:39     super(TestCase, self).assertLessEqual(abs(x - y), prec, message) 
Jan 03 00:55:39 AssertionError: 1024 not less than or equal to 1e-05 : __main__.TestAutogradDeviceTypeCUDA.test_logdet_1x1_cuda leaked 1024 bytes CUDA memory on device 0 
Jan 03 00:55:39  
Jan 03 00:55:39 ---------------------------------------------------------------------- 
Jan 03 00:55:39 Ran 1884 tests in 903.312s 
Jan 03 00:55:39  
Jan 03 00:55:39 FAILED (failures=1, skipped=13, expected failures=1) 
Jan 03 00:55:39  
Jan 03 00:55:39 Generating XML reports... 
Jan 03 00:55:40 Traceback (most recent call last): 
Jan 03 00:55:40   File "test/run_test.py", line 456, in <module> 
Jan 03 00:55:40     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: test/dist_autograd_test.py
Comment (fkhan1337) [reply test/dist_autograd_test.py]: I believe you meant `snake_case` - changed accordingly.

File: test/dist_autograd_test.py
Comment (fkhan1337) [reply test/dist_autograd_test.py]: The test seems to fail a couple of times when I did that. I'm not sure if it matters that `t1` is `ones` and `t2` is `zeros`.

**Review by rohan-varma — APPROVED** [review]
Looks good, thanks for working on this! Please make sure all tests pass before landing.

File: test/dist_autograd_test.py
Comment (rohan-varma) [reply test/dist_autograd_test.py]: At first glance this shouldn't happen but shouldn't block this PR either. I will investigate more later.

**Pull request closed without merging**
